site/src/pages/GroupsPage/GroupPage.tsx

```diff
@@ -39,76 +37,38 @@ import {
 } from "components/TableToolbar/TableToolbar";
 import { UserAvatar } from "components/UserAvatar/UserAvatar";
 import { isEveryoneGroup } from "utils/groups";
-
-const AddGroupMember: React.FC<{
-  isLoading: boolean;
-  onSubmit: (user: User, reset: () => void) => void;
-}> = ({ isLoading, onSubmit }) => {
-  const [selectedUser, setSelectedUser] = useState<User | null>(null);
-  const styles = useStyles();
-
-  const resetValues = () => {
-    setSelectedUser(null);
-  };
-
-  return (
-    <form
-      onSubmit={(e) => {
-        e.preventDefault();
-
-        if (selectedUser) {
-          onSubmit(selectedUser, resetValues);
-        }
-      }}
-    >
-      <Stack direction="row" alignItems="center" spacing={1}>
-        <UserAutocomplete
-          className={styles.autoComplete}
-          value={selectedUser}
-          onChange={(newValue) => {
-            setSelectedUser(newValue);
-          }}
-        />
-
-        <LoadingButton
-          disabled={!selectedUser}
-          type="submit"
-          startIcon={<PersonAdd />}
-          loading={isLoading}
-        >
-          Add user
-        </LoadingButton>
-      </Stack>
-    </form>
-  );
-};
+import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
+import {
+  addMember,
+  deleteGroup,
+  group,
+  groupPermissions,
+  removeMember,
+} from "api/queries/groups";
+import { displayError, displaySuccess } from "components/GlobalSnackbar/utils";
+import { getErrorMessage } from "api/errors";
 
 export const GroupPage: React.FC = () => {
-  const { groupId } = useParams();
-  if (!groupId) {
-    throw new Error("groupId is not defined.");
-  }
-
+  const { groupId } = useParams() as { groupId: string };
+  const queryClient = useQueryClient();
   const navigate = useNavigate();
-  const [state, send] = useMachine(groupMachine, {
-    context: {
-      groupId,
-    },
-    actions: {
-      redirectToGroups: () => {
-        navigate("/groups");
-      },
-    },
-  });
-  const { group, permissions } = state.context;
-  const isLoading = group === undefined || permissions === undefined;
+  const groupQuery = useQuery(group(groupId));
+  const groupData = groupQuery.data;
+  const { data: permissions } = useQuery(groupPermissions(groupId));
+  const addMemberMutation = useMutation(addMember(queryClient));
+  const removeMemberMutation = useMutation(removeMember(queryClient));
+  const deleteGroupMutation = useMutation(deleteGroup(queryClient));
+  const [isDeletingGroup, setIsDeletinGroup] = useState(false);
+  const isLoading = !groupData || !permissions;
```

Comment:
TS stuff only. Unfortunately, if I use `query.isLoading` conditional the nested children do not know the `query.data` is not undefined 😢 this is the only reason. But it has been awhile since the last time I checked this so I will try again to see if for some reason, with the new react-query and TS versions, it works now.

Reply:
Yeah, I think I've generally just used `onSuccess` to do the type narrowing, but I see the benefit here. This way, if `isLoading` is false, you do get a guarantee that the data exists